Introduce ExternalTypeRefInfoProvider

augusto2112 opened the pull request:
Introduce ExternalTypeRefInfoProvider

LLDB would like to cache typeref information to accelerate finding type
information. This patch adds an optional interface that allows of
typeref to register and provider field descriptor information for faster
lookups.

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -772,7 +772,7 @@ class ReflectionContext
   /// \return
```
comment is now out of date

adrian-prantl commented on include/swift/Remote/TypeInfoProvider.h:
```diff
@@ -32,6 +45,26 @@ struct TypeInfoProvider {
+  virtual void registerFieldDescriptors(
+      uint64_t InfoID,
+      const swift::reflection::FieldSection &FieldDescriptors,
+      const std::vector<std::string> &MangledNames) = 0;
```
`const std::vector` sounds like it could be `ArrayRef`?

adrian-prantl commented on include/swift/Remote/TypeInfoProvider.h:
```diff
@@ -32,6 +45,26 @@ struct TypeInfoProvider {
+      const std::vector<std::string> &MangledNames) = 0;
+
+  /// Retrieve a pair representing the reflection info id and the offset of a
+  /// field descriptor in the field section buffer, if available.
```
maybe define a struct with descriptive field names for the pair?

adrian-prantl commented on stdlib/public/Reflection/TypeRefBuilder.cpp:
```diff
@@ -263,23 +316,32 @@ RemoteRef<FieldDescriptor> TypeRefBuilder::getFieldTypeInfo(const TypeRef *TR) {
-          return *FD;
+      if ((!ExternalTypeInfo ||
+           ExternalTypeInfo->shouldParseReflectionInfoWithId(i)))
+        if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
```
Not sure if this is much better, bug have you considered a SmallSet for PotentialModuleNames (in the small case it's also a linear list).

augusto2112 replied:
This unfortunately can't be a SmallSet because the stdlib doesn't ship with it. I'm not even sure a set would be that much faster since there are so few elements.

adrian-prantl commented on include/swift/Remote/TypeInfoProvider.h:
```diff
@@ -32,6 +45,26 @@ struct TypeInfoProvider {
+
+  /// Register the field descriptors of a reflection info with a given id with
+  /// their corresponding mangled names. The amount of field descriptors and
+  /// mangled names should be the same. If a field descriptor does not have a
```
should or must?

augusto2112 commented:
@swift-ci smoke test

mikeash commented on include/swift/Reflection/TypeRefBuilder.h:
```diff
@@ -982,35 +991,46 @@ class TypeRefBuilder {
   IntVariableReader OpaqueIntVariableReader;
 
 public:
-  template<typename Runtime>
```
Looks like this is just a formatting change. Can we do that as a separate PR if so?

augusto2112 replied:
It got formatted because of the new `remote::ExternalTypeRefInfoProvider *provider = nullptr` parameter, but I guess it formatted the whole constructor 😅  I'll remove the non-related parts.

mikeash commented:
Neat idea. Would it make sense to roll this logic into Remote Mirror rather than having it provided externally, or is it going to be very specific to LLDB in some way?

It would also be good to have tests to exercise the case when an external provider is given.

augusto2112 commented:
> Neat idea. Would it make sense to roll this logic into Remote Mirror rather than having it provided externally, or is it going to be very specific to LLDB in some way?

Could you explain what you're envisioning when you suggest rolling this logic into Remote Mirror? If it could benefit the runtime besides LLDB I'd be happy to do it. To give more context, LLDB's use case for this is to cache a map of mangled names to field descriptor locators (the field descriptor and the offset in the typeref buffer), and then provide them back in future runs. You can check the lldb patch here: https://github.com/apple/llvm-project/pull/5100. The interesting file to look at i s is [SwiftMetadataCache.h](https://github.com/apple/llvm-project/pull/5100/files#diff-a88003491c8e8284dcf602fe7b2ba220aaecbda06c5ad5dada987607e0efd702R2] and [SwiftMetadataCache.cpp](https://github.com/apple/llvm-project/pull/5100/files#diff-9b59346d1112105e1631ae380eeb8b8c4e8fd4ea677bf0ec12ede7f1ddeee735R1]. If you see a way this could benefit the runtime as well we could change the approach.

augusto2112 committed:
Introduce ExternalTypeRefCache

LLDB would like to cache typeref information to accelerate finding type
information. This patch adds an optional interface that allows of
typeref to register and provider field descriptor information for faster
lookups.

augusto2112 commented:
> It would also be good to have tests to exercise the case when an external provider is given.

I'll try to come up with a way of testing this.

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test

mikeash commented:
@augusto2112 and I discussed this some offline. Persistent caching wouldn't fit well on the Swift side. It might be feasible to put most of the caching here, with a get/set function that the client could use to persist it across runs. Then clients other than LLDB could use it pretty easily if they needed it. However, we don't have such a need yet, so we'll just keep that as a future possibility.

adrian-prantl commented on stdlib/public/Reflection/TypeRefBuilder.cpp:
```diff
@@ -263,23 +309,32 @@ RemoteRef<FieldDescriptor> TypeRefBuilder::getFieldTypeInfo(const TypeRef *TR) {
-          return *FD;
+      if ((!ExternalProvider ||
+           ExternalProvider->shouldParseReflectionInfoWithId(i)))
+        if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
```
I think this could use another comment explaining the algorithm and the contract with ExternalProvider here. I'm not sure it's obvious to the reader that !shouldParseReflectionInfoWithId(i) means we already parsed it.

adrian-prantl commented on stdlib/public/Reflection/TypeRefBuilder.cpp:
```diff
@@ -263,23 +309,32 @@ RemoteRef<FieldDescriptor> TypeRefBuilder::getFieldTypeInfo(const TypeRef *TR) {
-    if (auto FD = findFieldDescriptorAtIndex(i, *MangledName))
-      return *FD;
+    if ((!ExternalProvider ||
+         ExternalProvider->shouldParseReflectionInfoWithId(i)))
```
would
```
if (ExternalProvider && !ExternalProvider->shouldParseReflectionInfoWithId(i)))
  continuel
```
be easier to read?

adrian-prantl replied:
"shouldParse" doesn't explain why. Should we name it "didCacheReflectionInfo(i)" or simething more descriptive?

augusto2112 replied:
I named the ExternalProvider in a deliberately not specific way just in case someone else wanted to implement the interface with a different strategy. I can change it to be more concrete though.

adrian-prantl commented on include/swift/Remote/ExternalTypeRefInfoProvider.h:
```diff
@@ -0,0 +1,74 @@
+
+  /// Returns whether looking for a mangled name in reflection info with a given
+  /// id could possibly return information that couldn't be provided by the
+  /// external provider.
```
What would be an example?

augusto2112 replied:
If we have all of the field descriptors for that ReflectionInfo on the cache, and `getFieldDescriptorLocator` didn't find anything, then the field descriptor we're looking for must not be in that ReflectionInfo.

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test

augusto2112 commented:
@adrian-prantl made the new interface closer to how it's actually implemented, which should make the logic easier to understand.

adrian-prantl commented on include/swift/Remote/ExternalTypeRefCacheProvider.h:
```diff
@@ -0,0 +1,74 @@
+//===--- ExternalTypeRefCacheProvider.h - Abstract access to external caches of
+//typeref ------*- C++ -*-===//
```
The `ExternalTypeInfoProvider` is called so because it provides typeinfo records from the outside. Here, I think `ExternalTypeRefCache` alone would suffice (without the `provider)?

adrian-prantl replied:
Thinking about it ... do we even need the `External` in the type name? I.e.: we could pass around a `TypeRefCache *ExternalTypeRefCache`

augusto2112 replied:
I agree with removing "provider", I do think "external" in the type name makes it more obvious what it is, versus as the variable name.

adrian-prantl commented on include/swift/Remote/ExternalTypeRefCacheProvider.h:
```diff
@@ -0,0 +1,74 @@
+//===----------------------------------------------------------------------===//
+//
+//  This file declares an abstract interface for external cache providers of
+//  typeref information
```
`.`

Also should this be a file-wide doxygen comment?

adrian-prantl commented on include/swift/Remote/ExternalTypeRefCacheProvider.h:
```diff
@@ -0,0 +1,74 @@
+}
+
+namespace remote {
+/// A struct with the information requited to locate a specific field
```
`required`

adrian-prantl commented on include/swift/Remote/ExternalTypeRefCacheProvider.h:
```diff
@@ -0,0 +1,74 @@
+  virtual void cacheFieldDescriptors(
+      uint64_t InfoID,
+      const swift::reflection::FieldSection &FieldDescriptors,
+      const llvm::ArrayRef<std::string> &MangledNames) = 0;
```
`llvm::ArrayRef` is meant to be used as a value type, no need for the `const&`

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test Linux

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -228,7 +229,8 @@ class ReflectionContext
 
   template <typename T>
-  bool readMachOSections(
+  llvm::Optional<uint64_t>
```
document what the return value is?

adrian-prantl replied:
And, like in the other patch — can this be a `uint32_t`?

adrian-prantl reviewed (approved):
LGTM with inline comments addressed

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test Linux

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5100

@swift-ci smoke test macOS
